### PR TITLE
Sample application to chat with NIMs and ai.nvidia.com

### PR DESCRIPTION
**Pull request by mocsharp**
A sample application that demonstrates the ability to use OpenAI APIs with NVIDIA Inference Microservices (NIMs) whether self-hosted or from build.nvidia.com.

chatwithnim

### CODE REVIEW

**Comment by jjomier** [comment]
I would recommend renaming that application to something more explicit: sample_nim?

**Comment by jjomier** [comment]
@mocsharp are we ready to merge?

File: applications/nvidia_nim/nvidia_nim_chat/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+Continue to the [Start the Application](#start-the-application) section once inside the Docker container.
+
+### Run the Application without Docker
+
+Install all dependencies from the `requirements.txt` file:
+
+```bash
+# optionally create a virtual environment and activate it
+python3 -m venv .venv
+source.venv/bin/activate
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/README.md:55]:
```suggestion
source .venv/bin/activate
```

File: applications/nvidia_nim/nvidia_nim_chat/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+# install the required packages
+pushd applications/nvidia_nim/chat
+pip install -r requirements.txt
+popd
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/README.md:60]: Nitpick:

```suggestion
pip install -r applications/nvidia_nim/chat/requirements.txt
```

File: applications/nvidia_nim/nvidia_nim_chat/README.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+### Start the Application
+
+To use the NIMs on [build.nvidia.com/](https://build.nvidia.com/), configure your API key in the `nvidia_nim.yaml` configuration file and run the sample app as follows:
+
+note: you may also configure your api key using an environment variable.
+E.g., `export OPENAI_API_KEY=...`
+
+```bash
+# To use NVIDIA hosted NIMs available on build.nvidia.com, export your API key first
+export OPENAI_API_KEY=[enter your api key here]
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/README.md:72]: I'm not clear, does `OPENAI_API_KEY` need to be available as an environment variable to use NVIDIA hosted NIMs or is it sufficient to add the API key to `nvidia_nim.yaml`?

File: applications/nvidia_nim/nvidia_nim_chat/README.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+cd applications/nvidia_nim/chat
+python3 app.py
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/README.md:75]:
```suggestion
./run launch nvidia_nim_chat
```

File: applications/nvidia_nim/nvidia_nim_chat/README.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+## Connecting with Locally Hosted NIMs
+
+To use a locally hosted NIM, first download and start the NIM.
+Then configure the `host` parameter in the `nvidia_nim.yaml` configuration file to point to your local NIM instance.
+
+For example:
+
+```bash
+nim:
+  base_url: http://0.0.0.0:8000/v1/
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/README.md:90]: Should the user configure `host`, `base_url`, or both in the YAML file?

File: applications/nvidia_nim/nvidia_nim_chat/README.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+models:
+  llama3-8b-instruct:
+    model: meta-llama3-8b-instruct
+    top_p: 1
+    n: 1
+    max_tokens: 1024
+    frequency_penalty: 1.0
+    stream: True
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/README.md:99]: It's not clear how this section relates to the local NIM configuration. Are we forwarding arbitrary arguments to a local NIM?

If not, suggest removing this block.

```suggestion
```

File: applications/nvidia_nim/nvidia_nim_chat/metadata.json [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+		"language": "Python",
+		"version": "1.0",
+		"changelog": {
+			"1.0": "Initial Release"
+		},
+		"holoscan_sdk": {
+			"minimum_required_version": "1.0.0",
+			"tested_versions": [
+				"1.0.0"
+			]
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/metadata.json:19]: Can you confirm the tested version? Looks like 2.1.0 based on `requirements.txt`:
```suggestion
			"minimum_required_version": "2.1.0",
			"tested_versions": [
				"2.1.0"
			]
```

File: applications/nvidia_nim/nvidia_nim_chat/metadata.json [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+		],
+		"tags": [
+			"OpenAI API",
+			"NIM"
+		],
+		"ranking": 1,
+		"dependencies": {},
+		"run": {
+			"command": "python3 <holohub_app_source>/app.py",
+			"workdir": "holohub_app_bin"
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/metadata.json:33]: Given that files do not seem to be copied into the binary directory, we probably want to run from the source directory:

```suggestion
			"workdir": "holohub_app_source"
```

File: applications/nvidia_nim/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+RUN echo ". /etc/bash_completion.d/holohub_autocomplete" >> /etc/bash.bashrc
+
+# - This variable is consumed by all dependencies below as an environment variable (CMake 3.22+)
+# - We use ARG to only set it at docker build time, so it does not affect cmake builds
+#   performed at docker run time in case users want to use a different BUILD_TYPE
+ARG CMAKE_BUILD_TYPE=Release
+
+
+COPY applications/nvidia_nim/chat/requirements.txt /tmp/requirements.txt
+RUN pip install -r /tmp/requirements.txt --no-cache-dir
```
Comment (tbirdso) [inline applications/nvidia_nim/Dockerfile:52]: Suggest moving this file one folder down to `nvidia_nim_chat` given that it install dependencies specific to the chat application.

File: applications/nvidia_nim/nvidia_nim_chat/README.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+The `models` section int the YAML file is configured with multiple NVIDIA hosted models by default. This allows you to switch between different models easily within the application by sending `/m` to the application.
+
+Model parameters may be added or adjusted in the `models` section as well per model.
+
+## Run the sample application
+
+There are a couple of options to run the sample application:
+
+### Run using Docker
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/README.md:31]: Suggest "quick start" instructions with `dev_container build_and_run`:
1. Add API key in `nvidia_nim.yaml`
2. `./dev_container build_and_run nvidia_nim_chat`

Expected result is that the Dockerfile is detected, container builds, app builds, and app launches automatically with one command for streamlined user experience.

File: applications/nvidia_nim/nvidia_nim_chat/README.md (hunk 3 above)
Comment (mocsharp) [reply applications/nvidia_nim/nvidia_nim_chat/README.md:72]: Let me rename that to API_KEY

File: applications/nvidia_nim/nvidia_nim_chat/metadata.json (hunk 7 above)
Comment (mocsharp) [reply applications/nvidia_nim/nvidia_nim_chat/metadata.json:19]: This app should work with 1.0. I will confirm.

File: applications/nvidia_nim/nvidia_nim_chat/README.md (hunk 5 above)
Comment (mocsharp) [reply applications/nvidia_nim/nvidia_nim_chat/README.md:87]: Thanks. I renamed it but forgot to update it here.

File: applications/nvidia_nim/nvidia_nim_chat/README.md (hunk 10 above)
Comment (mocsharp) [reply applications/nvidia_nim/nvidia_nim_chat/README.md:36]: Added Quick Start to the top of the README. Thank you!

File: applications/nvidia_nim/nvidia_nim_chat/README.md (hunk 6 above)
Comment (mocsharp) [reply applications/nvidia_nim/nvidia_nim_chat/README.md]: I've added comment to the example to explain what these are.

File: applications/nvidia_nim/nvidia_nim_chat/metadata.json (hunk 8 above)
Comment (mocsharp) [reply applications/nvidia_nim/nvidia_nim_chat/metadata.json]: Thank you.

File: applications/nvidia_nim/nvidia_nim_chat/metadata.json (hunk 7 above)
Comment (mocsharp) [reply applications/nvidia_nim/nvidia_nim_chat/metadata.json]: I have tested the app working using 1.0.3 of the PyPI wheel. Thanks!

File: applications/nvidia_nim/nvidia_nim_chat/metadata.json [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+		"language": "Python",
+		"version": "1.0",
+		"changelog": {
+			"1.0": "Initial Release"
+		},
+		"holoscan_sdk": {
+			"minimum_required_version": "1.0.3",
+			"tested_versions": [
+				"1.0.3"
+			]
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/metadata.json:19]: The `./dev_container build` or `./dev_container build_and_run` options currently use v2.1.0 by default. Could you please do a quick test to verify that the app runs with v2.1.0 as well? Sufficient to confirm `./dev_container build_and_run nvidia_nim_chat` runs as expected.

```suggestion
			"tested_versions": [
				"1.0.3",
				"2.1.0"
			]
```

Comment (tbirdso) [reply applications/nvidia_nim/nvidia_nim_chat/metadata.json:19]: Alternatively, if you do want to pin to 1.0.3 you can add the` --base_img` flag to those `./dev_container` commands in the app README to use the 1.0.3 base container.

Comment (mocsharp) [reply applications/nvidia_nim/nvidia_nim_chat/metadata.json]: `./dev_container build_and_run nvidia_nim_chat` works as expected. `2.1.0` has been added to `tested_versions`.

File: applications/nvidia_nim/nvidia_nim_chat/requirements.txt [hunk 12]
```diff
@@ -0,0 +1,3 @@
+halo==0.0.31
+holoscan==1.0.3
```
Comment (tbirdso) [inline applications/nvidia_nim/nvidia_nim_chat/requirements.txt:2]: Nitpick: could we pin to the latest tested version (v2.1.0)?
```suggestion
holoscan==2.1.0
```

**Review by tbirdso — APPROVED** [review]
One version nitpick, otherwise latest changes look good to me. Thanks!

**Pull request merged**